lib/rebaser-server/src/middleware.rs

```diff
@@ -0,0 +1,153 @@
+use naxum::{
+    async_trait, composite_rejection, define_rejection,
+    extract::{FromMessage, FromMessageHead},
+    Head, MessageHead,
+};
+use rebaser_core::{
+    api_types::{ApiVersionsWrapper, ApiWrapper},
+    nats,
+};
+use si_data_nats::Subject;
+use telemetry::prelude::*;
+
+define_rejection! {
+    #[status_code = 400]
+    #[body = "Invalid decoding string as utf8"]
+    pub struct InvalidUtf8Error(Error);
+}
+
+#[derive(Debug)]
+pub struct HeaderReply(pub Option<Subject>);
+
+#[async_trait]
+impl<S> FromMessageHead<S> for HeaderReply {
+    type Rejection = InvalidUtf8Error;
+
+    async fn from_message_head(head: &mut Head, _state: &S) -> Result<Self, Self::Rejection> {
+        let maybe_value = head.headers.as_ref().and_then(|headers| {
+            headers
+                .get(nats::NATS_HEADER_REPLY_INBOX_NAME)
+                .map(|value| value.to_string())
+        });
+
+        match maybe_value {
+            Some(value) => Ok(Self(Some(
+                Subject::from_utf8(value).map_err(InvalidUtf8Error::from_err)?,
+            ))),
+            None => Ok(Self(None)),
+        }
+    }
+}
+
+define_rejection! {
+    #[status_code = 400]
+    #[body = "Headers are required for content info but none was found"]
+    pub struct HeadersMissing;
+}
+
+define_rejection! {
+    #[status_code = 400]
+    #[body = "Failed to parse content info from headers"]
+    pub struct HeadersParseError(Error);
+}
+
+composite_rejection! {
+    pub enum ContentInfoRejection {
+        HeadersMissing,
+        HeadersParseError,
+    }
+}
+
+#[derive(Debug)]
+pub struct ContentInfo(pub rebaser_core::ContentInfo<'static>);
+
+#[async_trait]
+impl<S> FromMessageHead<S> for ContentInfo {
+    type Rejection = ContentInfoRejection;
+
+    async fn from_message_head(head: &mut Head, _state: &S) -> Result<Self, Self::Rejection> {
+        let headers = head.headers.as_ref().ok_or(HeadersMissing)?;
+        let content_info =
+            rebaser_core::ContentInfo::try_from(headers).map_err(HeadersParseError::from_err)?;
+
+        Ok(Self(content_info))
+    }
+}
+
+define_rejection! {
+    #[status_code = 400]
+    #[body = "failed to deserialize message payload"]
+    pub struct DeserializeError(Error);
+}
+
+define_rejection! {
+    #[status_code = 400]
+    #[body = "failed to upgrade message to current version"]
+    pub struct MessageUpgradeError(Error);
+}
+
+define_rejection! {
+    #[status_code = 415]
+    #[body = "unsupported content type"]
+    pub struct UnsupportedContentTypeError;
+}
+
+define_rejection! {
+    #[status_code = 406]
+    #[body = "unsupported message type"]
+    pub struct UnsupportedMessageTypeError;
+}
+
+define_rejection! {
+    #[status_code = 406]
+    #[body = "unsupported message version"]
+    pub struct UnsupportedMessageVersionError;
+}
```

Comment:
These are clever. Nice.

Reply:
A port over from tricks in the Axum code base. Yes, I like these macros!